worker/upgradeseries/worker.go

```diff
@@ -328,6 +341,86 @@ func (w *upgradeSeriesWorker) compareUnitAgentServices(units []names.UnitTag) (m
 	return unitServices, true, nil
 }
 
+// unpinLeaders pins leadership for applications based on unit agent services
+// running on the machine.
+func (w *upgradeSeriesWorker) pinLeaders() (err error) {
+	apps, err := w.applications()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// if we encounter an error from here on,
+	// attempt to reverse any completed leader pinning.
+	defer func() {
+		if err != nil {
+			if unpinErr := w.unpinLeaders(); unpinErr != nil {
+				err = errors.Wrap(err, unpinErr)
+			}
+		}
+	}()
+	for _, app := range apps {
+		if err := w.PinLeadership(app); err != nil {
+			return errors.Trace(err)
+		}
+		w.logger.Infof("pinned leader for application %q", app)
+	}
+
+	w.leadersPinned = true
+	return nil
+}
+
+// unpinLeaders unpins leadership for applications based on unit agent services
+// running on the machine.
+func (w *upgradeSeriesWorker) unpinLeaders() error {
+	apps, err := w.applications()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// We attempt to unpin for all applications,
+	// returning only the last error.
+	for _, app := range apps {
+		err = w.UnpinLeadership(app)
+		if err == nil {
+			w.logger.Infof("unpinned leader for application %q", app)
+		}
+	}
+
+	if err == nil {
+		w.leadersPinned = false
+	}
+	return errors.Trace(err)
+}
+
+// applications returns a collection of application names from the
+// unit agent services running on the machine.
+func (w *upgradeSeriesWorker) applications() ([]string, error) {
```

Comment:
machine.Applications() could almost be pushed all the way down into State as it seems a fairly generic request. It would also help if we decided to restrict permissions to only those applications on a machine.

Reply:
Yes, good point. I'll make a new method for this purpose.